### PR TITLE
Add Heroku deploy with Terraform addon

### PR DESCRIPTION
**Pull request by namtx**
## What happened
In this PR I introduce Terraform addon which helps deploy the application to Heroku.

## Insight
- https://registry.terraform.io/providers/heroku/heroku/latest/docs
- https://devcenter.heroku.com/articles/scaling
## Proof Of Work
- We can successfully deploy the application to Heroku 🚀 

rails-template-test-heroku___Heroku

With properly **Config Vars** 🛠️ 

rails-template-test-heroku_·_Settings___Heroku

Load Balancing and Scaling with [Heroku Formation](https://devcenter.heroku.com/articles/scaling) 📈 

rails-template-test-heroku_·_Resources___Heroku

Application logs

tmux_new_-snimblehq

### CODE REVIEW

**Comment by andyduong1920** [comment]
@namtx The test is failing 🚨

**Comment by namtx** [comment]
@andyduong1920 there is a Semaphore's networking error, it's 🆗 now

File: .template/addons/terraform/heroku/config.tf.tt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+resource "heroku_config" "common" {
+  vars = {
+    RAILS_LOG_TO_STDOUT      = "true"
+    RAILS_SERVE_STATIC_FILES = "true"
+    MAILER_DEFAULT_HOST      = "localhost"
+    MAILER_DEFAULT_PORT      = "3000"
```
Comment (andyduong1920) [inline .template/addons/terraform/heroku/config.tf.tt:6]: 3000 isn't MAILER_PORT I think @namtx

File: .template/addons/terraform/heroku/heroku.yml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+build:
+  docker:
+    web: Dockerfile
+run:
+  web: bin/start.sh
```
Comment (andyduong1920) [inline .template/addons/terraform/heroku/heroku.yml:5]: How about the Worker? Do we need to start the worker? @namtx 🤔

File: Makefile [hunk 3]
```diff
@@ -10,7 +10,7 @@ run:
 	cd $(APP_NAME) && \
 	docker-compose -f docker-compose.test.yml run test
 
-base_spec = spec/base/**/*_spec.rb,spec/addons/docker/**/*_spec.rb
+base_spec = spec/base/**/*_spec.rb,spec/addons/docker/**/*_spec.rb,spec/addons/terraform/**/*_spec.rb
```
Comment (andyduong1920) [inline Makefile:13]: The Makefile was refactored in #248, so please rebase this PR @namtx

File: Makefile [hunk 4]
```diff
@@ -1,5 +1,5 @@
 create:
-	rails new $(APP_NAME) -m ./template.rb -T ${OPTIONS}
+	printf "Y\n" | rails new $(APP_NAME) -m ./template.rb -T ${OPTIONS}
```
Comment (andyduong1920) [inline Makefile:2]: Please add the comment what does the `Y` do here @namtx, you can check how other addon did that.

File: .template/addons/terraform/heroku/variables.tf.tt [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+variable "redis_plan" {
+  description = "Heroku Redis addon plan"
+  default     = "hobby-dev" # Free
+}
+
+variable "dyno_size" {
+  description = "Heroku Formation dyno size"
+  default     = "Standard-1x"
```
Comment (andyduong1920) [inline .template/addons/terraform/heroku/variables.tf.tt:35]: should we set it as free as default :D

File: .template/addons/terraform/heroku/variables.tf.tt [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+variable "dyno_quantity" {
+  description = "Heroku Formation dyno quantity"
+  default     = 2
```
Comment (andyduong1920) [inline .template/addons/terraform/heroku/variables.tf.tt:40]: should we set 1 as default?

File: .template/addons/terraform/heroku/variables.tf.tt (hunk 5 above)
Comment (namtx) [reply .template/addons/terraform/heroku/variables.tf.tt]: changed it to `Free`.

File: Makefile [hunk 7]
```diff
@@ -13,6 +13,7 @@ test_variant_app:
 base_addon_spec = spec/addons/base/**/*_spec.rb
 web_addon_spec = spec/addons/variants/web/**/*_spec.rb
 api_addon_spec = spec/addons/variants/api/**/*_spec.rb
+terraform_addon_spec = spec/addons/terraform/**/*_spec.rb
```
Comment (andyduong1920) [inline Makefile:16]: I don't think we need this as the `terraform` could be located inside the base addon folder 🤔

**Comment by andyduong1920** [comment]
@namtx should we consider adjust the `.gitignore` and `.dockerignore` to ignore some part from Terraform, example the `.terraform` folder 🤔

File: .template/addons/terraform/heroku/variables.tf.tt (hunk 6 above)
Comment (namtx) [reply .template/addons/terraform/heroku/variables.tf.tt]: set it to 1, thanks for pointing

File: Makefile (hunk 3 above)
Comment (namtx) [reply Makefile]: rebased and made some changes already, thanks for pointing

File: .template/addons/terraform/heroku/heroku.yml (hunk 2 above)
Comment (namtx) [reply .template/addons/terraform/heroku/heroku.yml:6]: good point, `worker` is already added.

File: .template/addons/terraform/heroku/config.tf.tt (hunk 1 above)
Comment (namtx) [reply .template/addons/terraform/heroku/config.tf.tt]: changed it to 587 already

**Comment by ankitkalia1195** [comment]
Closed, at it is required to have infra in seperate repo as per our new discussion.

**Pull request closed without merging**